lxd/storage/drivers/lock.go

```diff
@@ -17,9 +18,14 @@ var lxdStorageOngoingOperationMap = map[string]chan bool{}
 // lxdStorageMapLock is used to access lxdStorageOngoingOperationMap.
 var lxdStorageMapLock sync.Mutex
 
-func lock(lockID string) func() {
+// Lock creates a lock with the given lockID. On success, it returns an unlock
```

Comment:
@monstermunchkin the comment here is out of date, suggest something long the lines of "Lock creates a lock for a specific storage volume to allow activities that require exclusive access to take place. Will block until the lock is established."

Reply:
@monstermunchkin as I was looking at your other commits that used this Lock function I noticed that you were having to check for whether the `Lock()` function returned nil each time. This got me thinking "why can that Lock function fail?".

So taking a closer look at the lock function below I wondered why the lxdStorageOngoingOperationMap is a map of wait channels rather than a map of `sync.Mutex`?

I was thinking that this same behaviour could be achieved by:

1. Calling `lxdStorageMapLock.Lock()` (as you do today).
2. Check whether the lockID entry exists in the `lxdStorageOngoingOperationMap` (as you do today).
3. But rather than the existing item being a channel if it were a `sync.Mutex` you could then call `lxdStorageMapLock.Unlock()` (as you do today).
4. If it doesn't exist then you'd create a new sync.Mutex item in the map.
5. Then you would call `item.Lock()` on the map item (which is either the existing lock or the new one you just created).
6. This would then lock until the lock is gained and which point you can return the unlock function that does everything it does today, except that it would call `item.Unlock()` rather than `close(waitChannel)`.

That way the Lock function could never fail, and you wouldn't need to test its return value every time you used it.

P.s. there's a typo on this line https://github.com/lxc/lxd/pull/6466/commits/2635ed99d7a1fb1ddd2d029bc4c2b216a22a8199#diff-519a50929b63143d5eb172b779a78a47L28 but if you switch to use sync.Mutex it will go anyway.

Reply:
@tomponline Do you mean like this? This will actually cause a deadlock.

```go
var lxdStorageOngoingOperationMap = map[string]*sync.Mutex{}
var lxdStorageMapLock sync.Mutex

func Lock(poolName string, volType VolumeType, volName string) func() {
	lockID := fmt.Sprintf("%s/%s/%s", poolName, volType, volName)

	lxdStorageMapLock.Lock()

	mutex, ok := lxdStorageOngoingOperationMap[lockID]
	if !ok {
		mutex = &sync.Mutex{}
		lxdStorageOngoingOperationMap[lockID] = mutex
	}

	lxdStorageMapLock.Unlock()

	mutex.Lock()

	return func() {
		lxdStorageMapLock.Lock()

		mutex, ok := lxdStorageOngoingOperationMap[lockID]
		if ok {
			mutex.Unlock()
			delete(lxdStorageOngoingOperationMap, lockID)
		}

		lxdStorageMapLock.Unlock()
	}
}
```

Reply:
As we chatted about on IRC, this one should do the trick https://play.golang.org/p/2kIkuXm6URQ